torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp

```diff
@@ -1729,16 +1735,30 @@ void ProcessGroupNCCL::heartbeatMonitor() {
   if (checkDumpSignal && shouldDump_.load()) {
     // Store debug info to storage if no other thread does it. (By default to
     // local disk)
-    std::future<bool> asyncDebugDump = std::async(
+    std::future<bool> asyncDebugDumpWithStack = std::async(
         std::launch::async, [this]() { return this->dumpDebuggingInfo(); });
 
     // wait for the dump until timeout - log data
-    waitForFutureOrTimeout(
-        asyncDebugDump,
+    auto result = waitForFutureOrTimeout(
+        asyncDebugDumpWithStack,
         std::chrono::milliseconds(waitTimeoutDumpInMilSec_),
         "Flight recorder dump in heartbeatMonitor",
         false,
         true);
+
+    // If we failed to dump, try dumping without stack trace.
+    if (!result) {
+      std::future<bool> asyncDebugDumpWithNoStack =
+          std::async(std::launch::async, [this]() {
+            return this->dumpDebuggingInfo(/*captureStackTrace=*/false);
+          });
+      waitForFutureOrTimeout(
+          asyncDebugDumpWithNoStack,
+          std::chrono::milliseconds(waitTimeoutDumpInMilSec_),
+          "Flight recorder dump in heartbeatMonitor",
+          false,
+          true);
+    }
```

Comment:
nit: maybe consider sharing code between the true and false cases and put them in a `do-while` loop?

Reply:
Good suggestion:
Used the following to reuse code.
```
    bool captureStackTrace = true;
    for (int i = 0; i < 2; i++) {
      std::future<bool> asyncDebugDump = std::async(
          std::launch::async, [this, captureStackTrace]() { return this->dumpDebuggingInfo(captureStackTrace); });

      // wait for the dump until timeout - log data
      auto result = waitForFutureOrTimeout(
          asyncDebugDump,
          std::chrono::milliseconds(waitTimeoutDumpInMilSec_),
          "Flight recorder dump in heartbeatMonitor",
          false,
          true);

      if (result) {
        break;        
      }
      // If we failed to dump, try dumping without stack trace.
      captureStackTrace = false;
    }
```